File: shared/typescript/src/agent-turn.ts
/**
 * AgentTurn request/response — the primary agent-api contract.
 * Mirrors ../contracts/agent-turn.schema.json.
 * Used on POST /v1/agent/turn, called by echokit-server on every final STT transcript.
 */

import { z } from "zod";
import {
  ConversationTurnSchema,
  IntentSchema,
  SessionEntitiesSchema,
} from "./session.js";
import { VehicleSchema } from "./vehicle.js";

export const ActionSchema = z.enum([
  "respond",
  "transfer",
  "collect_lead",
  "end_call",
]);
export type Action = z.infer<typeof ActionSchema>;

export const RagSourceSchema = z.enum([
  "faq",
  "policy",
  "promo",
  "financing",
  "service",
]);
export type RagSource = z.infer<typeof RagSourceSchema>;

export const RagChunkSchema = z
  .object({
    id: z.string().nullable().optional(),
    source: RagSourceSchema,
    content: z.string(),
    score: z.number().min(0).max(1),
  })
  .strict();
export type RagChunk = z.infer<typeof RagChunkSchema>;

export const LatencyMetricsSchema = z
  .object({
    intent_ms: z.number().min(0).default(0),
    rag_ms: z.number().min(0).default(0),
    sql_ms: z.number().min(0).default(0),
    synthesis_ms: z.number().min(0).default(0),
    total_ms: z.number().min(0).default(0),
  })
  .strict();
export type LatencyMetrics = z.infer<typeof LatencyMetricsSchema>;

export const AgentTurnResultsSchema = z
  .object({
    rag_chunks: z.array(RagChunkSchema).default([]),
    vehicles: z.array(VehicleSchema).default([]),
  })
  .strict();
export type AgentTurnResults = z.infer<typeof AgentTurnResultsSchema>;

export const AgentTurnRequestSchema = z
  .object({
    call_uuid: z.string().min(1),
    transcript: z.string(),
    caller_phone_hash: z.string(),
    session_entities: SessionEntitiesSchema.default({ features: [] }),
    conversation_history: z.array(ConversationTurnSchema).default([]),
    stt_confidence: z.number().min(0).max(1).nullable().optional(),
  })
  .strict();
export type AgentTurnRequest = z.infer<typeof AgentTurnRequestSchema>;

// E.164: + followed by 8-15 digits, first digit non-zero
const E164_REGEX = /^\+[1-9][0-9]{7,14}$/;

export const AgentTurnResponseSchema = z
  .object({
    response_text: z.string(),
    intent: IntentSchema,
    intent_confidence: z.number().min(0).max(1).default(0),
    updated_entities: SessionEntitiesSchema.default({ features: [] }),
    results: AgentTurnResultsSchema.default({ rag_chunks: [], vehicles: [] }),
    action: ActionSchema,
    transfer_to: z.string().regex(E164_REGEX).nullable().optional(),
    latency_ms: LatencyMetricsSchema.default({
      intent_ms: 0,
      rag_ms: 0,
      sql_ms: 0,
      synthesis_ms: 0,
      total_ms: 0,
    }),
  })
  .strict();
export type AgentTurnResponse = z.infer<typeof AgentTurnResponseSchema>;

File: shared/python/src/voiceai_contracts/agent_turn.py
"""AgentTurn request/response — the primary agent-api contract.

Mirrors shared/contracts/agent-turn.schema.json.
Used on POST /v1/agent/turn — called by echokit-server on every final STT transcript.
"""

from __future__ import annotations

from enum import Enum

from pydantic import BaseModel, ConfigDict, Field

from voiceai_contracts.session import ConversationTurn, Intent, SessionEntities
from voiceai_contracts.vehicle import Vehicle


class Action(str, Enum):
    RESPOND = "respond"
    TRANSFER = "transfer"
    COLLECT_LEAD = "collect_lead"
    END_CALL = "end_call"


class RagSource(str, Enum):
    FAQ = "faq"
    POLICY = "policy"
    PROMO = "promo"
    FINANCING = "financing"
    SERVICE = "service"


class RagChunk(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str | None = None
    source: RagSource
    content: str
    score: float = Field(ge=0.0, le=1.0)


class LatencyMetrics(BaseModel):
    model_config = ConfigDict(extra="forbid")

    intent_ms: float = Field(default=0.0, ge=0.0)
    rag_ms: float = Field(default=0.0, ge=0.0)
    sql_ms: float = Field(default=0.0, ge=0.0)
    synthesis_ms: float = Field(default=0.0, ge=0.0)
    total_ms: float = Field(default=0.0, ge=0.0)


class AgentTurnResults(BaseModel):
    model_config = ConfigDict(extra="forbid")

    rag_chunks: list[RagChunk] = Field(default_factory=list)
    vehicles: list[Vehicle] = Field(default_factory=list)


class AgentTurnRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    call_uuid: str = Field(min_length=1)
    transcript: str
    caller_phone_hash: str
    session_entities: SessionEntities = Field(default_factory=SessionEntities)
    conversation_history: list[ConversationTurn] = Field(default_factory=list)
    stt_confidence: float | None = Field(default=None, ge=0.0, le=1.0)


class AgentTurnResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    response_text: str
    intent: Intent
    intent_confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    updated_entities: SessionEntities = Field(default_factory=SessionEntities)
    results: AgentTurnResults = Field(default_factory=AgentTurnResults)
    action: Action
    transfer_to: str | None = Field(default=None, pattern=r"^\+[1-9][0-9]{7,14}$")
    latency_ms: LatencyMetrics = Field(default_factory=LatencyMetrics)
